### ISSUE TITLE
Check lock

### ISSUE THREAD

**Issue by Arthur111** [mattmassicotte/Lock#3]
Hello,

Thank you for you project, it fill the gap Apple did with async and NSLock.

Sometimes we need to check lock without changing the status, so maybe this function can be added : 

``` 
public func isLocked() -> Bool
{
  switch state
  {
    case .unlocked: return false
    case .locked:   return true
  }
}
```

When unlock but it was already unlocked, just do nothing (avoid fatal error), so maybe changed like this : 

```
public func unlock()
{
  switch state
  {
    case .unlocked: return
    case .locked:   state.resumeNextContinuation()
  }
}
```

**Comment by mattmassicotte** [comment]
Hello!

Thanks for both the kinds words and the bug report - super apperciated!

I totally agree about the `isLocked` API. Just added it!

I'm much more hesitant about making `unlock` silently fail when the lock is actually unlocked. Can you think of a situation where this makes sense? I think in all cases, this should be a programming error. The `withLock` function can help to make this more convenient. Perhaps that will work?

**Comment by Arthur111** [comment]
> Can you think of a situation where this makes sense?

Yes with client server application with many tests, process, error management (catch), cleanup function.
NSLock already behaves in this way.
Maybe put an option "check" with default set to "false".
By default this will change nothing to the actual behaviour.

**Comment by mattmassicotte** [comment]
I still don't think I understand. A `defer` or a `withLock`, both of which are extremely common patterns for locking, should be able to deal with all of these, no?

Now, the exception could be cleanup function. I have not thought hard about this. Can this be implemented safely?

**Comment by Arthur111** [comment]
NSLock already behaves in this way.

````
        let a = NSLock()
        
        a.lock()
        
        a.unlock()
        a.unlock()
````

There is no error.

For example when a communication is closed, interrupted (for whatever reason) between a client and a server, you take action on this event (reinit unclock), and you don't know if there was a lock or not.

**Comment by Arthur111** [comment]
Something like that : 

````
    public func unlock(check: Bool = false)
    {
        switch state
        {
            case .unlocked: if !check { state.resumeNextContinuation() }
            case .locked:   state.resumeNextContinuation()
        }
    }
````

**Comment by mattmassicotte** [comment]
Right, I understand the functionality and the implementation. What I mean is, why is this good API? Because it permits an unmatched `unlock` it also makes it very easy to accidetally mis a needed lock. So, I just need to understand why that risk is worth it.

For example, these are both much safer constructs:

```swift
let a = NSLock()
defer { a.unlock() }
// no code paths can now forget to unlock

a.withLock {
  // do stuff that requires the lock
}
```

**Comment by Arthur111** [comment]
Ok. i will explain better.

Client server connection.

The client, user send a file (or data) --> Lock to protect the user to send another file before the first one is received.
The server receive the file and send a ok that the file is received.
The client receive the ok and unlock.
Ok ?
Now if the communication is interrupted (broken or closed by the user)  how do i know the user is sending the file or not ? I don't know if there is a lock or not.
There is no mismatch, at first.
It is the interruption of communication who can create a mismatch.
I don't have a variable set to "sending" yes or no.
So i have to unlock and reinit some data because the communication is closed.
Instead of checking "if a.isLocked { a.unlock }",  i think it is better to do -> unlock(check: true) 

i can't use withLock because there are different called and calling functions. (it is a client/server communication, dialogue).

**Comment by mattmassicotte** [comment]
Ah ha! I didn't realize you were holding the lock across a multi-step transaction. You totally could do this, but this library is made for dealing with re-entrancy exclusively.

Adding this kind of API opens up class of bug that currently cannot happen. I *think* (but please correct me), that if you do want to use the lock state as a flag you can now do so with just a little more code right?

**Comment by Arthur111** [comment]
The lock is used to wait the transaction is complete to process the next one. (if there is one). 
The lock guarantee the whole transaction is ok and complete.
The transaction being the dialogue between the client and the server.
It worked well with NSLock but NSLock does not work with async / await.
It work well with your program (except the unlock behaves differently than NSLock).

**Comment by Arthur111** [comment]
--> fatalError("Continuations cannot be added when unlocked")

Are you sure its the same message for both "addContinuation" and "resumeNextContinuation" and not "Continuations cannot be removed when unlocked" for "resumeNextContinuation" ?

**Comment by mattmassicotte** [comment]
Ohh thank you! fixed! b5432fbbb53ed21341d77a28fdcba8e312288653

**Comment by mattmassicotte** [comment]
As for changing the `unlock` behavior. I thought more about this, and consulted with some others.

The goal here isn't to match `NSLock` behavior at all. It is to help deal with actor re-entrancy. And I believe the type, as is, can do this. Given that there is now a `isLocked` property, locking the actor across method calls can still be done. You can also achieve a similar thing with plain old actor state.

So, all that said, I think I'm happy with the functionality as-is. Hopefully I haven't disappointed too much!
